[READY]Custom reagent icecream + tweaks

Krysonism opened the pull request:
![milkies](https://user-images.githubusercontent.com/49783092/60549442-244eb700-9d25-11e9-8a8f-a0c235571b2c.png)
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

This pull requests lets the ice cream vat dispense custom ice creams flavoured with a reagent of your choice from an internal beaker. The ice cream takes the name and reagent colour of the master reagent.

You can also easily refill the ice cream vats ingredient storage by inserting a beaker and clicking the refill from beaker button.

I also plan on adjusting the colour of a few reagent that i think deserve good looking ice creams but do not have the colour for it.

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game

The ice cream vat is currently really lame, the fun comes from making strange and interesting flavours, a feature that is absent from /tg/. No more!

The ice creams in this screenshot are flavoured with earthsblood, crank, romerol, krokodil, turbo and happiness, but the possibilities are endless, you can be bold and incorporate some mayonnaise or just play it safe with caramel.

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: NT has made breakthroughs in ice cream science, ice creams can now be flavoured with any reagent!
tweak: The ice cream vat now accepts beakers.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Krysonism committed:
Merge remote-tracking branch 'upstream/master'

Krysonism committed:
Merge pull request #3 from tgstation/master

update fork

Krysonism committed:
Merge pull request #5 from tgstation/master

update master

Krysonism committed:
Merge pull request #7 from tgstation/master

update fork again

Krysonism committed:
Merge pull request #9 from tgstation/master

update fork

Krysonism committed:
Merge pull request #10 from tgstation/master

update fork

Krysonism committed:
The icecream vat now accepts beakers 

And could theoretically refill from them if i can get this to work.

Krysonism committed:
Custom reagent icecreams!

Custom reagent icecreams can now be scooped!

TerraGS commented:
Liquid Gibs ice cream

qwertyquerty commented on code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm:
```diff
@@ -211,6 +252,21 @@
 			desc = "A suspicious [cone_type] cone filled with bright red ice cream. That's probably not strawberry..."
 			reagents.add_reagent(/datum/reagent/liquidgibs, 2)
+		if ("custom")
+			if(R && R.total_volume >= 2)
```
`R?.total_volume >= 2`

Krysonism replied:
I will fix this today

qwertyquerty commented on code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm:
```diff
@@ -76,15 +82,22 @@
 	for(var/datum/reagent/R in reagents.reagent_list)
 		dat += "[R.name]: [R.volume]"
-		dat += "<A href='?src=[REF(src)];disposeI=[R.type]'>Purge</A><BR>"
+		dat += "<A href='?src=[REF(src)];disposeI=[R.type]'>Purge</a><BR>"
```
You randomly changed `</A>` to `</a>` here which doesn't match the previous tag.

Krysonism replied:
I must have done it trying to get dispose to  work. Any idea why the dispose button is broken?

It was broken before i started messing with the vat...

qwertyquerty replied:
Took a look and I'm not sure

ExcessiveUseOfCobblestone replied:
it still doesn't work after text2path?

ExcessiveUseOfCobblestone replied:
```suggestion
		dat += "<A href='?src=[REF(src)];disposeI=[R.type]'>Purge</A><BR>"
```

qwertyquerty commented on code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm:
```diff
@@ -222,10 +278,28 @@
+	if(!istype(user) || !user.canUseTopic(src, BE_CLOSE, FALSE, NO_TK))
+		return
+	replace_beaker(user)
+	return
```
Not sure the purpose of this return

Krysonism replied:
I will purge it

ExcessiveUseOfCobblestone replied:
```suggestion
```

qwertyquerty commented on code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm:
```diff
@@ -18,6 +19,7 @@
 	var/list/product_types = list()
 	var/dispense_flavour = ICECREAM_VANILLA
 	var/flavour_name = "vanilla"
+	var/obj/item/reagent_containers/beaker = null
```
I believe you are allowing any reagent container to work not just beakers. Keep in mind a lot of things are classified as reagent containers.

GuyonBroadway replied:
This means you can make pizza flavoured icecream, you tell me that isn't a hilarious notion.

Krysonism replied:
I don't think the machine will accept any non-open containers such as food, pills or patches.

qwertyquerty replied:
Ah I see now

ExcessiveUseOfCobblestone replied:
the last `/` is the var name so the `beaker` var is typed as `/obj/item/reagent_containers`

GuyonBroadway commented:
>That picture is named milkies.png

On a more serious note if you are going to borrow the goonstation feature of reagent icecream consider also borrowing the goonstation idea of the bottles to cheap drinks/condiments that contain a random assortment reagents that range from chilli juice to ants.

necromanceranne commented:
Meth icecream for a meth icecream cake with meth icing was how I killed a heavily armed traitor in hand to hand combat on goon once. He kept asking how I was still alive froma combination of baton smacks and a bullet lodged in me and I screamed, in my horrible swedish accent, METH CAKE.

Anyway I look forward to combining with botany to produce more meth ice cream cakes.

Krysonism commented:
> 
> 
> Meth icecream for a meth icecream cake with meth icing was how I killed a heavily armed traitor in hand to hand combat on goon once. He kept asking how I was still alive froma combination of baton smacks and a bullet lodged in me and I screamed, in my horrible swedish accent, METH CAKE.
> 
> Anyway I look forward to combining with botany to produce more meth ice cream cakes on tg as well. I've been wanting to see this feature for ages! It actually gives a good reason for chefs to reach out to their coworkers to produce cool food. Like the bartender, botany and chemistry.

Good point, i will try to make sure the ice cream transfers the correct colour to whichever custom food incorporates it.

By the way, currently the ice cream vat only transfers 2u of the reagent to the icecream to be consistent with the other flavour types, should we increase this somehow to 3 or 4, or should we just let the chef handle it by`manual syringe injection?

Krysonism commented:
> 
> 
> Liquid Gibs ice cream

I will probably make liquid gibs a deeper red colour since it makes a muddy orange icecream right now.

GuyonBroadway commented:
Clf3 ice cream should be called "spice-cream"

Krysonism committed:
Adds filling_color to the ice creams

Ice cream now properly colour custom foods.

Krysonism committed:
Vanilla ice cream now contain vanilla

Krysonism committed:
Ice creams are no longer bitter and have 50% more flavouring.

Krysonism committed:
Changes reagent colour of gibs and caramel.

Krysonism committed:
The ice creams are now very slightly more sugary and correct food types

Also the base reagents of icecream are now handled in a more sensible way.

OverlyGeneric commented:
we goon now boys. 

now we need randomized maintenance ice cream

necromanceranne commented:
Maybe have it be a setting in the ice cream vat for how many units of a reagent is used? 2u is likely not to do much with the fast metabolism reagents/toxins but that might be a good thing.

Krysonism commented:
> 
> 
> Maybe have it be a setting in the ice cream vat for how many units of a reagent is used? 2u is likely not to do much with the fast metabolism reagents/toxins but that might be a good thing.

I have increased it to 3u. 

If the chef desires to poison his ice cream he can inject 10u additional reagents with a syringe.

I might do the button ID later.

Krysonism committed:
Makes icecream making require more reagents.

This change brings the reagent cost of creating more ice cream in line with other desserts, also increases the bite size slightly to facilitate drugging / poisoning.

Krysonism commented:
Has anyone an idea why the dispose button is not working?

Krysonism committed:
Fixes the dispose bug

Thanks to AnturK on discord i was able to fix this bug.

Krysonism committed:
Changes the reagent content to make ice cream taste_mult = 1 friendly.

The custom ice creams now contain more flavouring and slightly less sugar so that you are able to taste reagents with taste_mult = 1.

Krysonism committed:
Caramel now tastes like caramel

Caramel has the taste "bitter sweetness" but since caramel is almost always going to be applied to sugary food its prone to producing taste messaged like "you can taste the sweetness and bitter sweetness", which are a bit repetitive. I have chosen to change the flavor to just "caramel" to fix this.

Krysonism commented:
I will be leaving for vacation in Germany for a week so i might not be able to make any edits until next week.

WaterStBlues commented:
Ice icecream

redmoogle commented:
I want this

Krysonism committed:
Update code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm

Co-Authored-By: moo <11748095+ExcessiveUseOfCobblestone@users.noreply.github.com>

Krysonism committed:
Update code/modules/food_and_drinks/kitchen_machinery/icecream_vat.dm

Co-Authored-By: moo <11748095+ExcessiveUseOfCobblestone@users.noreply.github.com>

Krysonism commented:
I am back in Sweden and ready address any suggested changes.
